ActorFrameworkIntegrationTest causes SIGSEG

**Summary**

> Try to answer the following as best as possible

- How often does the test fail? Once so far
- Does it block your work? not sure
- Do we suspect that it is a real failure? kind of

**Failures**

> Outline known failure cases, e.g. a failed assertion and its stacktrace obtained from Jenkins

<details><summary>Example assertion failure</summary>
<pre>
[INFO] Tests run: 6, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.018 s - in io.camunda.zeebe.dispatcher.AtomicPositionTest
[INFO] Running io.camunda.zeebe.dispatcher.integration.ActorFrameworkIntegrationTest
Aborted (core dumped)
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 21, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
Warning:  Corrupted channel by directly writing to native stream in forked JVM 1. See FAQ web page and the dump file /home/runner/work/zeebe/zeebe/dispatcher/target/surefire-reports/2022-09-20T07-11-50_009-jvmRun1.dumpstream
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  24.263 s
[INFO] Finished at: 2022-09-20T07:12:12Z
[INFO] ------------------------------------------------------------------------
Error:  Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.0.0-M7:test (default-test) on project zeebe-dispatcher: 
Error:  
Error:  Please refer to /home/runner/work/zeebe/zeebe/dispatcher/target/surefire-reports for the individual test results.
Error:  Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
Error:  The forked VM terminated without properly saying goodbye. VM crash or System.exit called?
Error:  Command was /bin/sh -c cd '/home/runner/work/zeebe/zeebe/dispatcher' && '/usr/lib/jvm/temurin-17-jdk-amd64/bin/java' '-javaagent:/home/runner/.m2/repository/org/jacoco/org.jacoco.agent/0.8.8/org.jacoco.agent-0.8.8-runtime.jar=destfile=/home/runner/work/zeebe/zeebe/dispatcher/target/jacoco.exec' '-jar' '/home/runner/work/zeebe/zeebe/dispatcher/target/surefire/surefirebooter-20220920071152080_3.jar' '/home/runner/work/zeebe/zeebe/dispatcher/target/surefire' '2022-09-20T07-11-50_009-jvmRun1' 'surefire-20220920071152080_1tmp' 'surefire_0-20220920071152080_2tmp'
Error:  Error occurred in starting fork, check output in log
Error:  Process Exit Code: 134
Error:  Crashed tests:
Error:  io.camunda.zeebe.dispatcher.integration.ActorFrameworkIntegrationTest
Error:  org.apache.maven.surefire.booter.SurefireBooterForkException: The forked VM terminated without properly saying goodbye. VM crash or System.exit called?
Error:  Command was /bin/sh -c cd '/home/runner/work/zeebe/zeebe/dispatcher' && '/usr/lib/jvm/temurin-17-jdk-amd64/bin/java' '-javaagent:/home/runner/.m2/repository/org/jacoco/org.jacoco.agent/0.8.8/org.jacoco.agent-0.8.8-runtime.jar=destfile=/home/runner/work/zeebe/zeebe/dispatcher/target/jacoco.exec' '-jar' '/home/runner/work/zeebe/zeebe/dispatcher/target/surefire/surefirebooter-20220920071152080_3.jar' '/home/runner/work/zeebe/zeebe/dispatcher/target/surefire' '2022-09-20T07-11-50_009-jvmRun1' 'surefire-20220920071152080_1tmp' 'surefire_0-20220920071152080_2tmp'
Error:  Error occurred in starting fork, check output in log
Error:  Process Exit Code: 134
Error:  Crashed tests:
Error:  io.camunda.zeebe.dispatcher.integration.ActorFrameworkIntegrationTest
Error:  	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.fork(ForkStarter.java:714)
Error:  	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.run(ForkStarter.java:311)
Error:  	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.run(ForkStarter.java:268)
Error:  	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.executeProvider(AbstractSurefireMojo.java:1334)
Error:  	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.executeAfterPreconditionsChecked(AbstractSurefireMojo.java:1167)
Error:  	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.execute(AbstractSurefireMojo.java:931)
Error:  	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:137)
Error:  	at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute(MojoExecutor.java:301)
Error:  	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:211)
Error:  	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:[165](https://github.com/camunda/zeebe/actions/runs/3088106271/jobs/4994204500#step:5:166))
Error:  	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:157)
Error:  	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:121)
Error:  	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:81)
Error:  	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:56)
Error:  	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:127)
Error:  	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:294)
Error:  	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:192)
Error:  	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:105)
Error:  	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:960)
Error:  	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:293)
Error:  	at org.apache.maven.cli.MavenCli.main(MavenCli.java:196)
Error:  	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
Error:  	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
Error:  	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
Error:  	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
Error:  	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:282)
Error:  	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:225)
Error:  	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:406)
Error:  	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:347)
Error:  -> [Help 1]
Error:  
Error:  To see the full stack trace of the errors, re-run Maven with the -e switch.
Error:  Re-run Maven using the -X switch to enable full debug logging.
Error:  
Error:  For more information about the errors and possible solutions, please read the following articles:
Error:  [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
Error: Process completed with exit code 1.
</pre>
</details>

**Hypotheses**

I think it is related to freeing again the internal dispatcher buffer, which was introduce with https://github.com/camunda/zeebe/pull/10381. The test might access still an buffer, which was already freed.

> List any hypotheses if you have one; can be ommitted

**Logs**


[hs_err_pid1970.log](https://github.com/camunda/zeebe/files/9605180/hs_err_pid1970.log)

--- a/dispatcher/src/test/java/io/camunda/zeebe/dispatcher/integration/ActorFrameworkIntegrationTest.java
+++ b/dispatcher/src/test/java/io/camunda/zeebe/dispatcher/integration/ActorFrameworkIntegrationTest.java
@@ -7,6 +7,8 @@
  */
 package io.camunda.zeebe.dispatcher.integration;
 
+import static org.assertj.core.api.Assertions.assertThat;
+
 import io.camunda.zeebe.dispatcher.BlockPeek;
 import io.camunda.zeebe.dispatcher.ClaimedFragment;
 import io.camunda.zeebe.dispatcher.Dispatcher;
@@ -18,10 +20,15 @@ import io.camunda.zeebe.scheduler.future.ActorFuture;
 import io.camunda.zeebe.scheduler.testing.ActorSchedulerRule;
 import io.camunda.zeebe.util.ByteValue;
 import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
 import org.agrona.DirectBuffer;
 import org.junit.Rule;
 import org.junit.Test;
 
+/**
+ * NOTE: make sure that all actors which will close over the dispatcher buffers are closed before
+ * closing the dispatcher, as that will free the underlying buffer.
+ */
 public final class ActorFrameworkIntegrationTest {
   @Rule public final ActorSchedulerRule actorSchedulerRule = new ActorSchedulerRule(3);
 
@@ -32,12 +39,15 @@ public final class ActorFrameworkIntegrationTest {
             .actorSchedulingService(actorSchedulerRule.get())
             .bufferSize((int) ByteValue.ofMegabytes(10))
             .build();
-
-    actorSchedulerRule.submitActor(new Consumer(dispatcher));
+    final Consumer consumer = new Consumer(dispatcher);
     final ClaimingProducer producer = new ClaimingProducer(dispatcher);
+
+    actorSchedulerRule.submitActor(consumer);
     actorSchedulerRule.submitActor(producer);
 
-    producer.latch.await();
+    assertThat(producer.latch.await(10, TimeUnit.SECONDS)).isTrue();
+    producer.close();
+    consumer.close();
     dispatcher.close();
   }
 
@@ -48,12 +58,15 @@ public final class ActorFrameworkIntegrationTest {
             .actorSchedulingService(actorSchedulerRule.get())
             .bufferSize((int) ByteValue.ofMegabytes(10))
             .build();
-
-    actorSchedulerRule.submitActor(new PeekingConsumer(dispatcher));
+    final PeekingConsumer consumer = new PeekingConsumer(dispatcher);
     final ClaimingProducer producer = new ClaimingProducer(dispatcher);
+
+    actorSchedulerRule.submitActor(consumer);
     actorSchedulerRule.submitActor(producer);
 
-    producer.latch.await();
+    assertThat(producer.latch.await(10, TimeUnit.SECONDS)).isTrue();
+    producer.close();
+    consumer.close();
     dispatcher.close();
   }
 

